tools/rewriter/GenCallAsm.cpp

```diff
@@ -481,21 +668,29 @@ static void emit_copy_stack_returns(AsmWriter &aw, size_t stack_return_size, siz
       add_comment_line(aw,
                        "Copy "s + std::to_string(stack_return_size) +
                            " bytes for the return value to the caller's stack");
-      for (int i = 0; i < stack_return_size; i += 8) {
-        add_asm_line(aw, "popq "s + std::to_string(i) + "(%rax)");
-      }
 
-      if (stack_return_padding > 0) {
-        add_asm_line(aw, "addq $8, %rsp");
-      }
+      emit_memcpy(aw, stack_return_size, "rax", "rsp", "r10", 0, arch);
+
+      add_asm_line(aw, "addq $"s + std::to_string(stack_return_size + stack_return_padding) + ", %rsp");
     }
   } else if (arch == Arch::Aarch64) {
-    // TODO ARM stack switch back
-    llvm::errs() << "TODO stack switch-back not implemented on ARM\n";
+    if (stack_return_size > 0) {
+      add_comment_line(aw, "Copy stack returns to caller's stack");
+
+      size_t stack_return_saved_ptr_offset = stack_arg_size + stack_arg_padding;
+      size_t stack_return_start = stack_return_saved_ptr_offset + 8;
+
+      // Pop the original return value address into x8
+      add_asm_line(aw, "ldr x8, [sp, #"s + std::to_string(stack_return_saved_ptr_offset) + "]");
+      // Copy the value back to the caller's stack
+      emit_memcpy(aw, stack_return_size, "x8", "sp", "x9", stack_return_start, arch);
```

Comment:
Since there's no intermediate PKRU we need to use tagged pointers for memory accesses right? I think all the other calls to emit_memcpy use pointers derived from sp, but I'm not totally sure about `x8` here since it's passed in as a hidden argument.

Reply:
I think we need to, yeah. I don't know how exactly we're doing this, maybe that should be a separate PR?

Reply:
Sure that can be a separate PR

Reply:
I imagine there might be more cases where we need to use tagged pointers in the call gates but we can handle those as a second pass in a separate PR. I'll just focus on whether it works w/o MTE for this PR.